// botbuilder-storage/lib/storage/MongoBotStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var crypto_1 = require("crypto");
var utils_1 = require("../utils");
var MongoBotStorage = (function () {
    function MongoBotStorage(mongoClient, settings) {
        this.mongoClient = mongoClient;
        this.settings = settings;
        var _a = settings || {}, collection = _a.collection, ttl = _a.ttl;
        if (!mongoClient || !collection) {
            throw new Error("Invalid constructor arguments for the MongoBotStorage class.");
        }
        if (ttl) {
            if (!utils_1.validateTTLSettings(ttl)) {
                throw new Error("Invalid TTL settings.");
            }
            else {
                this.ttl = ttl;
            }
        }
    }
    MongoBotStorage.prototype.getData = function (context, callback) {
        var readOps = [];
        var data = {};
        var userId = context.userId, conversationId = context.conversationId, persistUserData = context.persistUserData, persistConversationData = context.persistConversationData;
        if (userId) {
            if (persistUserData) {
                readOps.push({
                    _id: userId,
                    type: "userData",
                });
            }
            if (conversationId) {
                readOps.push({
                    _id: userId + ":" + conversationId,
                    type: "privateConversationData",
                });
            }
        }
        if (persistConversationData && conversationId) {
            readOps.push({
                _id: conversationId,
                type: "conversationData",
            });
        }
        var db = this.mongoClient;
        var c = this.settings.collection;
        Promise.all(readOps.map(function (entry) {
            return new Promise(function (resolve, reject) {
                var _id = entry._id, type = entry.type;
                db.collection(c).findOne({ _id: _id }, function (err, doc) {
                    if (err) {
                        return reject(err);
                    }
                    var docData = doc && doc.data || "{}";
                    var hash = doc && doc.hash;
                    var hashKey = type + "Hash";
                    data[type] = JSON.parse(docData);
                    data[hashKey] = hash;
                    resolve();
                });
            });
        })).then(function () {
            callback(null, data);
        }).catch(function (error) {
            callback(error, {});
        });
    };
    MongoBotStorage.prototype.saveData = function (context, data, callback) {
        var _this = this;
        var writeOps = [];
        var userId = context.userId, conversationId = context.conversationId, persistUserData = context.persistUserData, persistConversationData = context.persistConversationData;
        var addWrite = function (type, id, state, prevHash) {
            state = JSON.stringify(state || {});
            var hash = crypto_1.createHash("sha256").update(state);
            var newHash = hash.digest("hex");
            if (newHash !== prevHash) {
                var writeOperation = {
                    _id: id,
                    data: state,
                    hash: newHash,
                    type: type,
                    lastModified: new Date(),
                };
                if (_this.ttl) {
                    var timestamp = Date.now() + _this.ttl[type] * 1000;
                    writeOperation.expireAt = new Date(timestamp);
                }
                writeOps.push(writeOperation);
            }
        };
        addWrite = addWrite.bind(this);
        if (userId) {
            if (persistUserData) {
                addWrite("userData", userId, data.userData, data.userDataHash);
            }
            if (conversationId) {
                var id = userId + ":" + conversationId;
                var d = data.privateConversationData, h = data.privateConversationDataHash;
                addWrite("privateConversationData", id, d, h);
            }
        }
        if (persistConversationData && conversationId) {
            var d = data.conversationData, h = data.conversationDataHash;
            addWrite("conversationData", conversationId, d, h);
        }
        var db = this.mongoClient;
        var collection = this.settings.collection;
        Promise.all(writeOps.map(function (entry) {
            return new Promise(function (resolve, reject) {
                var _id = entry._id, data = entry.data, hash = entry.hash, type = entry.type, lastModified = entry.lastModified, expireAt = entry.expireAt;
                var doc = { data: data, hash: hash, type: type, lastModified: lastModified };
                if (expireAt) {
                    doc.expireAt = expireAt;
                }
                var options = { upsert: true };
                db.collection(collection).update({ _id: _id }, doc, options, function (err) {
                    if (err) {
                        return reject(err);
                    }
                    resolve();
                });
            });
        })).then(function () {
            callback(null);
        }).catch(function (error) {
            callback(error);
        });
    };
    return MongoBotStorage;
}());
exports.MongoBotStorage = MongoBotStorage;
